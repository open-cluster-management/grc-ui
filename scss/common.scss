/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/


@import 'node_modules/carbon-components/scss/globals/scss/styles';
@import 'node_modules/carbon-components/scss/globals/scss/_colors';
@import 'node_modules/carbon-components/scss/globals/scss/_helper-mixins';
@import 'header-carbon-v9-migration';
@import 'variables';
@import '_red-hat-font';

@include redhat-font-face;

html, body, div, span, applet, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, acronym, address, big, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strike, strong, sub, sup, tt, var, b, u, i, center, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption, footer, header, hgroup, menu, nav, output, ruby, section, summary, time, mark, audio, video {
  font-family: RedHatText;
}

.bx--label, .bx--search-input, .bx--overflow-menu-options__btn, .bx--pagination__text, .bx--text-input, .bx--tooltip--icon__top::before, .bx--tooltip--icon__top::after, .bx--tooltip--icon__bottom::before, .bx--tooltip--icon__bottom::after {
  font-family: RedHatText;
}

html {
  height: 100%;
}

body {
  min-height: 100%;
  height: 100%;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  .bx--data-table-v2--zebra tbody tr:nth-child(2n) {
    background-color: $color__white;
  }
  .bx--tooltip--shown {
    font-size: .875rem;
  }
}

.page-content-container {
  display: flex;
  flex-direction: column;
  padding: 1.5rem 5% 3.375rem;

  .bx--select--inline{
    svg.bx--select__arrow{
      position: absolute
    }
  }
}

.content-spinner {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.resource-list-filter {
  margin-top: -1.5rem;
  padding-bottom: 3rem;
}

.bx--loading-overlay {
  z-index: 10000; // need higher z-index with modals open
}

.expand-vertically {
  height: 100%;
}

.nested-expandable-row {
  display: flex;
  flex-direction: column;
  .nested-expandable-row-header {
    font-weight: bold;
    color: $color__blue-50;
  }
  .nested-expandable-row-content {
    margin-top: -2rem;
  }
}

.no-resource {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 32.375rem;
  height: 25.6875rem;
  color: $color__blue-90;
  padding: 1.25rem 2.5rem;
  background-color: transparent;
  box-shadow: none;
  border: none;
  margin: 3rem auto 1rem auto;

  .button-group {
    display: flex;
    .buttons {
      margin: 0 1rem;
      button:hover {
        .doc-link {
          color: white;
        }
      }
      .doc-link {
        text-decoration: none;
      }
    }
  }

  .no-resource-icon {
    width: 22.5rem;
    height: 12.5rem;
  }

  .no-resource-title {
    font-weight: 500;
    margin-bottom: 0.5625rem;
    font-size: 1.8125rem;
    text-align: center;
    color: $grc_color_blue;
  }

  .no-resource-detail {
    font-weight: 300;
    font-size: 1rem;
    margin-bottom: 1.3125rem;
    text-align: center;
    line-height: 1.2;
    color: $color__navy-gray-6;

    > a {
      font-weight: 700;
      margin-right: 0.3125rem;
      font-size: .875rem;
    }
  }
}

.yamlEditorContainer {
  #brace-editor {
    min-width: 31.25rem;
    min-height: 31.25rem
  }
}
