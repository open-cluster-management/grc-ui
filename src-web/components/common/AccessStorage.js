/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

// sessionStorage access section
export const deleteSessionState = (sessionName) => {
  sessionStorage.removeItem(sessionName)
}

export const getSessionState = (sessionName) => {
  let state = {}
  const savedState = sessionStorage.getItem(sessionName)
  if (savedState && typeof state ===  'object') {
    try {
      state = JSON.parse(savedState)
    } catch (e) {
      //
    }
  } else if (typeof state ===  'string') {
    state = savedState
  }
  return state
}

export const saveSessionState = (sessionName, state) => {
  if (typeof state === 'string') {
    sessionStorage.setItem(sessionName, state)
  } else {
    // Convert our sets to arrays for stringify
    const stateObject = {}
    for (const key in state) {
      if (typeof state[key] === 'object') { // set is also object
        stateObject[key] = Array.from(state[key])
      } else {
        stateObject[key] = state[key]
      }
    }
    sessionStorage.setItem(sessionName, JSON.stringify(stateObject))
  }
}

// simply replace session value on the given key
export const replaceSessionPair = (sessionName, key, value) => {
  const state = getSessionState(sessionName)
  if (key) {
    state[key] = value
  }
  saveSessionState(sessionName, state)
}

// append session value on the given key
export const addSessionPair = (sessionName, key, value) => {
  const state = getSessionState(sessionName)
  if (key) {
    state[key] ? state[key] = new Set(state[key]) : state[key] = new Set()
    state[key].add(value)
  }
  saveSessionState(sessionName, state)
}
