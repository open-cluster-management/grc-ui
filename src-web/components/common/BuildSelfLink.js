/* Copyright (c) 2021 Red Hat, Inc. */
'use strict'
import _ from 'lodash'

export const buildSelfLinK = (data, kind) => {
  const apiGroupVersion = _.get(data, 'raw.apiVersion')
  const namespace = _.get(data, 'metadata.namespace')
  const name = _.get(data, 'metadata.name')
  const selfLink = _.get(data, 'metadata.selfLink')
  if (apiGroupVersion && namespace && kind && name) {
    return `/apis/${apiGroupVersion}/namespaces/${namespace}/${kind}/${name}`
  } else if (selfLink) {
    return selfLink
  }
  return ''
}
