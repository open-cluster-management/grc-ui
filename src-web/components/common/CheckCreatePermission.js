/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

const checkCreatePermission = (userAccess) => {
  // if create policy permission on any ns, set flag to 1
  let createFlag = 0
  if (userAccess && typeof userAccess === 'object') {
    const policyKey = 'policy.open-cluster-management.io/policies'
    for (const singleNSAccess of userAccess) {
      const rules = singleNSAccess.rules
      if (Array.isArray(rules['*/*']) &&
              (rules['*/*'].includes('*') || rules['*/*'].includes('create'))) {
        createFlag = 1
        break
      }
      if (Array.isArray(rules[policyKey]) &&
              (rules[policyKey].includes('*') || rules[policyKey].includes('create'))) {
        createFlag = 1
        break
      }
    }
  }
  return createFlag
}

export default checkCreatePermission
