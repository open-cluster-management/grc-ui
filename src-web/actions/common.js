/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
import _ from 'lodash'

import GrcApolloClient from '../../lib/client/apollo-client'
import { formatExpandablePolicies } from '../components/common/FormatTableData'
import {
  TABLE_PAGE_CHANGE, TABLE_SEARCH, TABLE_SORT, RESOURCE_RECEIVE_SUCCESS,
  REQUEST_STATUS, RESOURCE_RECEIVE_FAILURE, RESOURCE_REQUEST, RESOURCE_ADD,
  RESOURCE_MODIFY, RESOURCE_DELETE, RESOURCE_MUTATE, RESOURCE_MUTATE_SUCCESS,
  RESOURCE_MUTATE_FAILURE, AVAILABLE_FILTER_UPDATE, ACTIVE_FILTER_UPDATE,
  SECONDARY_HEADER_UPDATE, MODAL_UPDATE, POST_REQUEST, POST_RECEIVE_SUCCESS,
  POST_RECEIVE_FAILURE, PUT_REQUEST, PUT_RECEIVE_SUCCESS, PUT_RECEIVE_FAILURE,
  PATCH_REQUEST, PATCH_RECEIVE_SUCCESS, PATCH_RECEIVE_FAILURE, DEL_REQUEST,
  DEL_RECEIVE_SUCCESS, DEL_RECEIVE_FAILURE, CLEAR_REQUEST_STATUS, RESOURCE_RESET,
} from './index'

export const changeTablePage = ({page, pageSize}, resourceType) => ({
  type: TABLE_PAGE_CHANGE,
  page,
  pageSize,
  resourceType
})

export const searchTable = (search, resourceType) => ({
  type: TABLE_SEARCH,
  search,
  resourceType
})

export const sortTable = (sortDirection, sortColumn, resourceType) => ({
  type: TABLE_SORT,
  sortDirection,
  sortColumn,
  resourceType
})

export const receiveResourceSuccess = (response, resourceType) => ({
  type: RESOURCE_RECEIVE_SUCCESS,
  status: REQUEST_STATUS.DONE,
  items: response.items,
  resourceVersion: _.get(response, 'metadata.resourceVersion'), //only supported on k8s resoruces
  resourceType
})

export const receiveResourceError = (err, resourceType) => ({
  type: RESOURCE_RECEIVE_FAILURE,
  status: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const requestResource = (resourceType) => ({
  type: RESOURCE_REQUEST,
  status: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const addResource = (item, resourceType) => ({
  type: RESOURCE_ADD,
  resourceType: item.kind || resourceType,
  item
})

export const modifyResource = (item, resourceType) => ({
  type: RESOURCE_MODIFY,
  resourceType: item.kind || resourceType,
  item
})

export const deleteResource = (item, resourceType) => ({
  type: RESOURCE_DELETE,
  resourceType: item.kind || resourceType,
  item
})

export const mutateResource = (resourceType, resourceName) => ({
  type: RESOURCE_MUTATE,
  resourceName,
  resourceType,
})

export const mutateResourceSuccess = (resourceType, resourceName) => ({
  type: RESOURCE_MUTATE_SUCCESS,
  resourceName,
  resourceType,
})

export const mutateResourceFailure = (resourceType, error) => ({
  type: RESOURCE_MUTATE_FAILURE,
  postStatus: REQUEST_STATUS.ERROR,
  err: { error },
  resourceType,
})

export const fetchResources = (resourceType, vars) => {
  return (dispatch) => {
    dispatch(requestResource(resourceType))
    return GrcApolloClient.get(resourceType, vars)
      .then(response => {
        if (response.errors) {
          return dispatch(receiveResourceError(response.errors[0], resourceType))
        }
        return dispatch(receiveResourceSuccess({items: _.cloneDeep(formatExpandablePolicies(response.data.items))}, resourceType))
      })
      .catch(err => dispatch(receiveResourceError(err, resourceType)))
  }
}

export const fetchSingleResource = (resourceType, args) => {
  console.log(resourceType)
  return (dispatch) => {
    dispatch(requestResource(resourceType))
    return GrcApolloClient.getResource(resourceType, args)
      .then(response => {
        if (response.errors) {
          return dispatch(receiveResourceError(response.errors[0], resourceType))
        }
        return dispatch(receiveResourceSuccess({items: _.cloneDeep(response.data)}, resourceType))
      })
      .catch(err => dispatch(receiveResourceError(err, resourceType)))
  }
}

export const fetchResource = (resourceType, namespace, name) => {
  return (dispatch) => {
    dispatch(requestResource(resourceType))
    return GrcApolloClient.getResource(resourceType, {namespace, name})
      .then(response => {
        if (response.errors) {
          return dispatch(receiveResourceError(response.errors[0], resourceType))
        }
        return dispatch(receiveResourceSuccess({items: _.cloneDeep(response.data.items)}, resourceType))
      })
      .catch(err => dispatch(receiveResourceError(err, resourceType)))
  }
}

export const editResource = (resourceType, namespace, name, body, selfLink, resourcePath) => (dispatch => {
  dispatch(putResource(resourceType))
  return GrcApolloClient.updateResource(resourceType.name, namespace, name, body, selfLink, resourcePath)
    .then(response => {
      if (response.errors) {
        return dispatch(receivePutError(response.errors[0], resourceType))
      } else {
        dispatch(updateModal({open: false, type: 'resource-edit'}))
      }
      dispatch(fetchResources(resourceType))
      return dispatch(receivePutResource(response, resourceType))
    })
})

export const disableResource = (resourceType, namespace, name, body, selfLink, resourcePath) => (dispatch => {
  dispatch(patchResource(resourceType))
  return GrcApolloClient.updateResource(resourceType.name, namespace, name, body, selfLink, resourcePath)
    .then(response => {
      if (response.errors) {
        return dispatch(receivePatchError(response.errors[0], resourceType))
      } else {
        dispatch(updateModal({open: false, type: 'resource-disable'}))
      }
      dispatch(fetchResources(resourceType))
      return dispatch(receivePatchResource(response, resourceType))
    })
})

export const enforcResource = (resourceType, namespace, name, body, selfLink, resourcePath) => (dispatch => {
  dispatch(patchResource(resourceType))
  return GrcApolloClient.updateResource(resourceType.name, namespace, name, body, selfLink, resourcePath)
    .then(response => {
      if (response.errors) {
        return dispatch(receivePatchError(response.errors[0], resourceType))
      } else {
        dispatch(updateModal({open: false, type: 'resource-enforce'}))
      }
      dispatch(fetchResources(resourceType))
      return dispatch(receivePatchResource(response, resourceType))
    })
})

export const removeResource = (resourceType, vars) => async dispatch => {
  dispatch(delResource(resourceType))
  try {
    const response = await GrcApolloClient.remove(vars)
    if (response.errors) {
      return dispatch(receiveDelError(response.errors, resourceType))
    } else {
      dispatch(receiveDelResource(response, resourceType, vars))
    }
    dispatch(fetchResources(resourceType))
  } catch (err) {
    return dispatch(receiveDelError(err, resourceType))
  }
}

export const updateActiveFilters = (activeFilters) => ({
  type: ACTIVE_FILTER_UPDATE,
  activeFilters,
})

export const updateAvailableFilters = (availableFilters) => ({
  type: AVAILABLE_FILTER_UPDATE,
  availableFilters
})

export const updateSecondaryHeader = (title, tabs, breadcrumbItems, links, description, information) => ({
  type: SECONDARY_HEADER_UPDATE,
  title,
  tabs,
  breadcrumbItems,
  links,
  description,
  information
})

export const updateModal = (data) => ({
  type: MODAL_UPDATE,
  data
})

export const postResource = (resourceType) => ({ // TODO: Consider renaming
  type: POST_REQUEST,
  postStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receivePostResource = (item, resourceType) => ({
  type: POST_RECEIVE_SUCCESS,
  postStatus: REQUEST_STATUS.DONE,
  resourceType: item.kind || resourceType,
  item
})

export const receivePostError = (err, resourceType) => ({
  type: POST_RECEIVE_FAILURE,
  postStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const putResource = (resourceType) => ({ // TODO: Consider renaming
  type: PUT_REQUEST,
  putStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receivePutResource = (item, resourceType) => {
  return ({
    type: PUT_RECEIVE_SUCCESS,
    putStatus: REQUEST_STATUS.DONE,
    resourceType: item.kind || resourceType,
    item
  })
}

export const receivePutError = (err, resourceType) => ({
  type: PUT_RECEIVE_FAILURE,
  putStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const patchResource = (resourceType) => ({ // TODO: Consider renaming
  type: PATCH_REQUEST,
  patchStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receivePatchResource = (item, resourceType) => {
  return ({
    type: PATCH_RECEIVE_SUCCESS,
    patchStatus: REQUEST_STATUS.DONE,
    resourceType: item.kind || resourceType,
    item
  })
}

export const receivePatchError = (err, resourceType) => ({
  type: PATCH_RECEIVE_FAILURE,
  patchStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const delResource = (resourceType) => ({ // TODO: Consider renaming
  type: DEL_REQUEST,
  delStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receiveDelResource = (item, resourceType, resource) => ({
  type: DEL_RECEIVE_SUCCESS,
  delStatus: REQUEST_STATUS.DONE,
  resourceType: item.kind || resourceType,
  item,
  resource
})

export const receiveDelError = (err, resourceType) => ({
  type: DEL_RECEIVE_FAILURE,
  delStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const clearRequestStatus = (resourceType) => ({
  type: CLEAR_REQUEST_STATUS,
  resourceType: resourceType
})

export const resetResource = (resourceType) => ({
  type: RESOURCE_RESET,
  resourceType: resourceType
})

export const createResources = (resourceType, resourceJson) => {
  return (dispatch) => {
    dispatch(mutateResource(resourceType))
    return GrcApolloClient.createResources(resourceJson)
      .then(result => {
        if (result.data.createResources.errors && result.data.createResources.errors.length > 0){
          dispatch(mutateResourceFailure(resourceType, result.data.createResources.errors[0]))
        } else {
          dispatch(mutateResourceSuccess(resourceType))
        }
        return result
      })
  }
}

export const createAndUpdateResources = (resourceTypes, createList, updateList) => {
  return (dispatch) => {
    resourceTypes.forEach((resourceType) => {
      dispatch(mutateResource(resourceType))
    })
    return GrcApolloClient.createAndUpdateResources(createList, updateList)
      .then(result => {
        const errors = {
          Policy: {
            resourceType: 'Policy',
            error: '',
          },
          PlacementRule: {
            resourceType: 'PlacementRule',
            error: '',
          },
          PlacementBinding: {
            resourceType: 'PlacementBinding',
            error: '',
          }
        }
        if (_.get(result, 'data.createAndUpdateResources.create.errors') &&
          _.get(result, 'data.createAndUpdateResources.create.errors.length') > 0){
          _.get(result, 'data.createAndUpdateResources.create.errors', []).forEach((error) => {
            errors[error.kind].error = error.message
          })
        }
        if (_.get(result, 'data.createAndUpdateResources.update.errors') &&
          _.get(result, 'data.createAndUpdateResources.update.errors.length') > 0){
          _.get(result, 'data.createAndUpdateResources.update.errors', []).forEach((error) => {
            errors[error.kind].error = error.message
          })
        }
        let errored = false
        Object.keys(errors).forEach((key) => {
          const resp = errors[key]
          if (resp.error !== '') {
            errored = true
            dispatch(mutateResourceFailure(resp.resourceType, { message: resp.error }))
          }
        })
        if (!errored) {
          resourceTypes.forEach((resourceType) => {
            dispatch(mutateResourceSuccess(resourceType))
          })
        }
        return result
      })
  }
}

export const createResource = (resourceType, variables) => {
  return (dispatch) => {
    dispatch(postResource(resourceType))
    return GrcApolloClient.createResource(resourceType, variables)
      .then(response => {
        if (response.errors) {
          return dispatch(receivePostError(response.errors[0], resourceType))
        }

        return dispatch(receivePostResource(_.cloneDeep(response.data.setHelmRepo), resourceType))
      })
      .catch(err => dispatch(receivePostError(err, resourceType)))
  }
}
