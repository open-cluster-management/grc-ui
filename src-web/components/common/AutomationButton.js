/* Copyright (c) 2021 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import msgs from '../../nls/platform.properties'
import TruncateText from '../../components/common/TruncateText'
import { updateModal } from '../../actions/common'
import { POLICY_AUTOMATIONS } from '../../utils/client/queries'
import { Query } from '@apollo/client/react/components'
import {
    AcmLaunchLink
} from '@open-cluster-management/ui-components'
import {
    Button,
    Spinner,
} from '@patternfly/react-core'

import '../../scss/resource-filterbar.scss'

class AutomationButton extends React.Component {
  constructor (props) {
    super(props)
  }

  render() {
    const { item, locale, onClickAutomation } = this.props

    return (
        <Query query={POLICY_AUTOMATIONS} variables={{ namespace: item.metadata.namespace }}>
        {( result ) => {
          const { data={policyAutomations: []} } = result
          const { loading } = result
          let found = false
          let automationName = ''
          data.policyAutomations.forEach((automation) => {
            if (automation.spec && automation.spec.policyRef === item.metadata.name) {
              found = true
              automationName = automation.metadata.name
            }
          })
          if (loading) {
            return <Spinner size='md' />
          }
          if (found) {
            return <AcmLaunchLink links={[
              {
                  id: `automationButton-${automationName}`,
                  text: <TruncateText maxCharacters={20} text={automationName} />,
                  onClick: () => {
                    onClickAutomation(item)
                  },
                  label: true,
              },
            ]}></AcmLaunchLink>
          }
          return (
            <Button
              component="a"
              variant="link"
              className="automationButton"
              onClick= {() => {
                onClickAutomation(item)
              }}
            >
              {msgs.get('table.actions.automation.configure', locale)}
            </Button>
          )
        }}
        </Query>
      )
  }
}

AutomationButton.propTypes = {
  item: PropTypes.object,
  locale: PropTypes.string,
  onClickAutomation: PropTypes.func,
}

const mapDispatchToProps = (dispatch) => {
  const resourceTypeAuto = {
    name: 'HCMCompliance',
  }
  return {
    onClickAutomation: (data) => {
      dispatch(updateModal(
        { open: true, type: 'resource-automation', resourceTypeAuto,
          label: {
            primaryBtn: `modal.automation-${resourceTypeAuto.name.toLowerCase()}.heading`,
            label: `modal.automation-${resourceTypeAuto.name.toLowerCase()}.label`,
            heading: `modal.automation-${resourceTypeAuto.name.toLowerCase()}.heading`
          },
          data: { kind: resourceTypeAuto.name, ...data }}))
    }
  }
}

export default connect(null, mapDispatchToProps)(AutomationButton)
