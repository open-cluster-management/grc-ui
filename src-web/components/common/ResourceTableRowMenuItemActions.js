/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import { updateModal } from '../../actions/common'
import config from '../../../server/lib/shared/config'
import _ from 'lodash'

export const resourceActions = (action, dispatch, resourceType, data, history) => {
  switch (action) {
  case 'table.actions.policy.details': {
    return history.push(`${config.contextPath}/all/${data.namespace}/${data.name}`)
  }
  case 'table.actions.policy.edit':
  case 'table.actions.edit': {
    return history.push(`${config.contextPath}/all/${data.namespace}/${data.name}/edit`)
  }
  case 'table.actions.launch.cluster':{
    const consoleURL = _.get(data, 'consoleURL')
    if (consoleURL) {
      window.open(`${consoleURL}`, '_blank')
    }
    return
  }
  case 'table.actions.compliance.remove':
  case 'table.actions.policy.remove':
  case 'table.actions.remove': {
    return dispatch(updateModal(
      { open: true, type: 'resource-remove', resourceType,
        label: {
          primaryBtn: `modal.remove-${resourceType.name.toLowerCase()}.heading`,
          label: `modal.remove-${resourceType.name.toLowerCase()}.label`,
          heading: `modal.remove-${resourceType.name.toLowerCase()}.heading`
        },
        data: { apiVersion: resourceType.api_version, kind: resourceType.name, ...data }}))
  }
  case 'table.actions.policy.disable':
  case 'table.actions.disable': {
    return dispatch(updateModal(
      { open: true, type: 'resource-disable', resourceType,
        label: {
          primaryBtn: `modal.disable-${resourceType.name.toLowerCase()}.heading`,
          label: `modal.disable-${resourceType.name.toLowerCase()}.label`,
          heading: `modal.disable-${resourceType.name.toLowerCase()}.heading`
        },
        data: { apiVersion: resourceType.api_version, kind: resourceType.name, ...data }}))
  }
  case 'table.actions.policy.enable':
  case 'table.actions.enable': {
    return dispatch(updateModal(
      { open: true, type: 'resource-enable', resourceType,
        label: {
          primaryBtn: `modal.enable-${resourceType.name.toLowerCase()}.heading`,
          label: `modal.enable-${resourceType.name.toLowerCase()}.label`,
          heading: `modal.enable-${resourceType.name.toLowerCase()}.heading`
        },
        data: { apiVersion: resourceType.api_version, kind: resourceType.name, ...data }}))
  }
  case 'table.actions.enforce': {
    return dispatch(updateModal(
      { open: true, type: 'resource-enforce', resourceType,
        label: {
          primaryBtn: `modal.enforce-${resourceType.name.toLowerCase()}.heading`,
          label: `modal.enforce-${resourceType.name.toLowerCase()}.label`,
          heading: `modal.enforce-${resourceType.name.toLowerCase()}.heading`
        },
        data: { apiVersion: resourceType.api_version, kind: resourceType.name, ...data }}))
  }
  case 'table.actions.inform': {
    return dispatch(updateModal(
      { open: true, type: 'resource-inform', resourceType,
        label: {
          primaryBtn: `modal.inform-${resourceType.name.toLowerCase()}.heading`,
          label: `modal.inform-${resourceType.name.toLowerCase()}.label`,
          heading: `modal.inform-${resourceType.name.toLowerCase()}.heading`
        },
        data: { apiVersion: resourceType.api_version, kind: resourceType.name, ...data }}))
  }
  case 'table.actions.automation.edit':
  default:
    return undefined
  }
}
