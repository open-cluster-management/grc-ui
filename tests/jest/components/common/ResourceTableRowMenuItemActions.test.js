/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */

'use strict'

import { resourceActions } from '../../../../src-web/components/common/ResourceTableRowMenuItemActions'

describe('ResourceTableRowMenuItemActions component table.actions.policy.edit', () => {
  it('renders as expected', () => {
    const action = 'table.actions.policy.edit'
    const resourceType = {
      'name': 'HCMCompliance',
      'list': 'HCMComplianceList'
    }
    const dispatch = jest.fn()
    expect(resourceActions(action, dispatch, resourceType, {}, true, {})).toMatchSnapshot()
  })
})

describe('ResourceTableRowMenuItemActions component table.actions.launch.cluster', () => {
  it('renders as expected', () => {
    const action = 'table.actions.launch.cluster'
    const resourceType = {
      'name': 'HCMPolicyCluster',
      'list': 'HCMPolicyClusterList'
    }
    window.open = jest.fn()
    const dispatch = jest.fn()
    expect(resourceActions(action, dispatch, resourceType, {}, true, {})).toMatchSnapshot()
    window.open.mockClear()
  })
})

describe('ResourceTableRowMenuItemActions component table.actions.policy.edit', () => {
  it('renders as expected', () => {
    const action = 'table.actions.policy.sidepanel'
    const resourceType = {
      'name': 'HCMCompliance',
      'list': 'HCMComplianceList'
    }
    const dispatch = jest.fn()
    expect(resourceActions(action, dispatch, resourceType, {}, true, {})).toMatchSnapshot()
  })
})

describe('ResourceTableRowMenuItemActions component table.actions.compliance.remove', () => {
  it('renders as expected', () => {
    const action = 'table.actions.compliance.remove'
    const resourceType = {
      'name': 'HCMCompliance',
      'list': 'HCMComplianceList'
    }
    const dispatch = jest.fn()
    expect(resourceActions(action, dispatch, resourceType, {}, true, {})).toMatchSnapshot()
  })
})

describe('ResourceTableRowMenuItemActions component table.actions.policy.remove', () => {
  it('renders as expected', () => {
    const action = 'table.actions.policy.remove'
    const resourceType = {
      'name': 'HCMCompliance',
      'list': 'HCMComplianceList'
    }
    const dispatch = jest.fn()
    expect(resourceActions(action, dispatch, resourceType, {}, true, {})).toMatchSnapshot()
  })
})

describe('ResourceTableRowMenuItemActions component table.actions.remove', () => {
  it('renders as expected', () => {
    const action = 'table.actions.remove'
    const resourceType = {
      'name': 'HCMCompliance',
      'list': 'HCMComplianceList'
    }
    const dispatch = jest.fn()
    expect(resourceActions(action, dispatch, resourceType, {}, true, {})).toMatchSnapshot()
  })
})
