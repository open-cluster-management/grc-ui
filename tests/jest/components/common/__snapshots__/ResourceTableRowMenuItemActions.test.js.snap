// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceTableRowMenuItemActions component table.actions.compliance.remove renders as expected 1`] = `
Array [
  Object {
    "data": Object {
      "data": Object {
        "apiVersion": undefined,
        "kind": "HCMCompliance",
      },
      "label": Object {
        "heading": "modal.remove-hcmcompliance.heading",
        "label": "modal.remove-hcmcompliance.label",
        "primaryBtn": "modal.remove-hcmcompliance.heading",
      },
      "open": true,
      "resourceType": Object {
        "name": "HCMCompliance",
        "query": "ALL_POLICIES",
      },
      "type": "resource-remove",
    },
    "type": "MODAL_UPDATE",
  },
]
`;

exports[`ResourceTableRowMenuItemActions component table.actions.compliance.remove table.actions.policy.remove is called as expected 1`] = `
Array [
  Object {
    "data": Object {
      "data": Object {
        "apiVersion": undefined,
        "kind": "HCMCompliance",
      },
      "label": Object {
        "heading": "modal.remove-hcmcompliance.heading",
        "label": "modal.remove-hcmcompliance.label",
        "primaryBtn": "modal.remove-hcmcompliance.heading",
      },
      "open": true,
      "resourceType": Object {
        "name": "HCMCompliance",
        "query": "ALL_POLICIES",
      },
      "type": "resource-remove",
    },
    "type": "MODAL_UPDATE",
  },
]
`;

exports[`ResourceTableRowMenuItemActions component table.actions.compliance.remove table.actions.remove is called as expected 1`] = `
Array [
  Object {
    "data": Object {
      "data": Object {
        "apiVersion": undefined,
        "kind": "HCMCompliance",
      },
      "label": Object {
        "heading": "modal.remove-hcmcompliance.heading",
        "label": "modal.remove-hcmcompliance.label",
        "primaryBtn": "modal.remove-hcmcompliance.heading",
      },
      "open": true,
      "resourceType": Object {
        "name": "HCMCompliance",
        "query": "ALL_POLICIES",
      },
      "type": "resource-remove",
    },
    "type": "MODAL_UPDATE",
  },
]
`;

exports[`ResourceTableRowMenuItemActions table.actions.edit is called as expected 1`] = `
Array [
  Object {
    "data": Object {
      "action": "put",
      "data": Object {
        "kind": "HCMCompliance",
      },
      "editorMode": "json",
      "label": Object {
        "heading": "modal.edit-hcmcompliance.heading",
        "label": "modal.edit-hcmcompliance.label",
        "primaryBtn": "modal.button.submit",
      },
      "open": true,
      "resourceType": Object {
        "name": "HCMCompliance",
        "query": "ALL_POLICIES",
      },
      "type": "resource-edit",
    },
    "type": "MODAL_UPDATE",
  },
]
`;

exports[`ResourceTableRowMenuItemActions table.actions.launch.cluster is called as expected 1`] = `
Array [
  "consoleURL",
  "_blank",
]
`;
